gh-106016: Support customizing of module attributes access with __setattr__/__delattr__ (PEP 726) (#108261)
python/cpython: Doc/reference/datamodel.rst

.. index::
   single: __getattr__ (module attribute)
   single: __setattr__ (module attribute)
   single: __delattr__ (module attribute)
   single: __dir__ (module attribute)
   single: __class__ (module attribute)

Special names ``__getattr__``, ``__setattr__``, ``__delattr__`` and ``__dir__``
can be also used to customize access to module attributes.

The ``__getattr__`` function at the module level
should accept one argument which is the name of an attribute and return the
computed value or raise an :exc:`AttributeError`. If an attribute is
not found on a module object through the normal lookup, i.e.
:meth:`object.__getattribute__`, then ``__getattr__`` is searched in
the module ``__dict__`` before raising an :exc:`AttributeError`. If found,
it is called with the attribute name and the result is returned.

> The ``__setattr__`` function should accept two arguments, respectively, the
> name of an attribute and the value to be assigned and return :const:`None` or
> raise an :exc:`AttributeError`.  If present, this function overrides the

Reviewer: It's surprising to say "return None", it's rare to write explicitly ``return None`` in Python. I suggest to omit it:

```suggestion
The ``__setattr__`` function should accept two arguments, respectively, the
name of an attribute and the value to be assigned. It can
raise an :exc:`AttributeError`.  If present, this function overrides the
```

standard :func:`setattr` behaviour on a module.  For example::

   def __setattr__(name, value):
       print(f'Setting {name} to {value}...')
       globals()[name] = value

> The ``__delattr__`` function should accept one argument which is the name of an
> attribute and return :const:`None` or raise an :exc:`AttributeError`.  If present,

Reviewer: Same remark:

```suggestion
The ``__delattr__`` function should accept one argument which is the name of an
attribute. It can raise an :exc:`AttributeError`.  If present,
```

this function overrides the standard :func:`delattr` behaviour on a module.

The ``__dir__`` function should accept no arguments, and return a sequence of
strings that represents the names accessible on module. If present, this
function overrides the standard :func:`dir` search on a module.

For a more fine grained customization of the module behavior, one can set the
``__class__`` attribute of a module object to a subclass of
:class:`types.ModuleType`. For example::

   import sys
   from types import ModuleType

   class CallableModule(ModuleType):
       def __call__(self, *args, **kwargs):

>            raise RuntimeError("NO-body expects the Spanish Inquisition!")

Reviewer: I don't get your spelling, is it deliberate?

```suggestion
           raise RuntimeError("Nobody expects the Spanish Inquisition!")
```

Author: Yep.  E.g. https://en.wikipedia.org/wiki/The_Spanish_Inquisition_(Monty_Python)

Reviewer: Oh ok ok, I tried but failed to find the reference :-)

   sys.modules[__name__].__class__ = CallableModule

.. note::

>    Defining module ``__getattr__``/``__setattr__``/``__delattr__`` and setting
>    module ``__class__`` only affect lookups made using the attribute access

Reviewer: ```suggestion
   Defining module ``__getattr__``, ``__setattr__``, ``__delattr__``, ``__dir__``
   and setting module ``__class__`` only affect lookups made using the attribute access
```
